### PR TITLE
Safe rotate upgrade marker file whenever it's written

### PR DESCRIPTION
**Pull request by ycombinator**
## What does this PR do?

This PR writes upgrade marker contents to a temporary file first, before safely rotating it into the upgrade marker file's actual path.

## Why is it important?

To prevent the upgrade marker file at the actual path from ever showing up as truncated or partially written.

## Related issues

- Relates https://github.com/elastic/elastic-agent/issues/3947
- Follow up to https://github.com/elastic/elastic-agent/pull/3948#discussion_r1435184050

### CODE REVIEW

File: internal/pkg/agent/application/upgrade/marker_access_common.go [hunk 1]
```diff
@@ -7,10 +7,14 @@ package upgrade
 import (
 	"fmt"
 	"os"
+
+	"github.com/elastic/elastic-agent-libs/file"
 )
 
 func writeMarkerFileCommon(markerFile string, markerBytes []byte, shouldFsync bool) error {
-	f, err := os.OpenFile(markerFile, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
+	tmpMarkerFile := fmt.Sprintf("%s.tmp", markerFile)
```
Comment (cmacknz) [inline internal/pkg/agent/application/upgrade/marker_access_common.go:15]: Only one writer at a time is possible, so we don't need to worry about collisions in this file name?

**Review by cmacknz — APPROVED** [review]

**Comment by cmacknz** [comment]
The test added in https://github.com/elastic/elastic-agent/pull/3948/files covers the core functionality changed here.

File: internal/pkg/agent/application/upgrade/marker_access_common.go (hunk 1 above)
Comment (ycombinator) [reply internal/pkg/agent/application/upgrade/marker_access_common.go:15]: I'll add PID and timestamp to the temporary filename just to be safe.

Comment (ycombinator) [reply internal/pkg/agent/application/upgrade/marker_access_common.go]: Added in 09e6028a28.

Comment (cmacknz) [reply internal/pkg/agent/application/upgrade/marker_access_common.go]: Thanks!

**Comment by ycombinator** [comment]
> The test added in https://github.com/elastic/elastic-agent/pull/3948/files covers the core functionality changed here.

I enhanced that test in that PR (see 5af070e884f9df92df03e968cedd3017fa1677a8) to try and check that the upgrade marker file is never truncated whenever it's written.

**Comment by ycombinator** [comment]
The `make check-ci` failure is happening on `main` as well; created https://github.com/elastic/elastic-agent/pull/3956 to fix it.

**Comment by cmacknz** [comment]
Buildkite seems to only have known failures now. I could force merge this but I'm not going to do it without an OK from @ycombinator there aren't more changes coming.

**Comment by cmacknz** [comment]
Nevermind there are more relevant failures if I look at the actual unit tests:

```

=== FAIL: internal/pkg/agent/application/upgrade TestWriteMarkerFileWithTruncation (8.59s)
--
  | 2023-12-22 12:51:30 EST | marker_access_test.go:58:
  | 2023-12-22 12:51:30 EST | Error Trace:	C:/buildkite-agent/builds/bk-agent-prod-gcp-1703266071480677783/elastic/elastic-agent/internal/pkg/agent/application/upgrade/marker_access_test.go:58
  | 2023-12-22 12:51:30 EST | Error:      	Received unexpected error:
  | 2023-12-22 12:51:30 EST | failed to write upgrade marker file [C:\Users\buildkite\AppData\Local\Temp\TestWriteMarkerFileWithTruncation305396899\001\.update-marker] despite retrying: failed to safe rotate upgrade marker file: rename C:\Users\buildkite\AppData\Local\Temp\TestWriteMarkerFileWithTruncation305396899\001\.update-marker-4040-1703266986890.tmp C:\Users\buildkite\AppData\Local\Temp\TestWriteMarkerFileWithTruncation305396899\001\.update-marker: The process cannot access the file because it is being used by another process.
  | 2023-12-22 12:51:30 EST | Test:       	TestWriteMarkerFileWithTruncation


```

File: internal/pkg/agent/application/upgrade/marker_access_common.go [hunk 2]
```diff
@@ -7,10 +7,15 @@ package upgrade
...
 	"fmt"
 	"os"
+	"time"
+
+	"github.com/elastic/elastic-agent-libs/file"
 )
 
 func writeMarkerFileCommon(markerFile string, markerBytes []byte, shouldFsync bool) error {
-	f, err := os.OpenFile(markerFile, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
+	tmpMarkerFile := fmt.Sprintf("%s-%d-%d.tmp", markerFile, os.Getpid(), time.Now().UnixMilli())
```
Comment (leehinman) [inline internal/pkg/agent/application/upgrade/marker_access_common.go:16]: Can we just use: https://pkg.go.dev/os#CreateTemp instead?

**Review by leehinman — CHANGES_REQUESTED** [review]

**Comment by pierrehilbert** [comment]
`C:\Users\buildkite\AppData\Local\Temp\TestWriteMarkerFile93440394\001\.update-marker: The process cannot access the file because it is being used by another process.`

**Review by leehinman — APPROVED** [review]
LGTM

**Pull request merged**